### ISSUE TITLE
LLVM: Internal error: ELEMENT outside of range: 2 

### ISSUE THREAD

**Issue by waywardmonkeys** [dylan-lang/opendylan#893]
I'm not sure what I broke yet that causes this in my code, but I can get this error:

```
Internal error: ELEMENT outside of range: 2
Exiting with return code -1
```

The stack trace at the time:

```
(lldb) dylan-bt
  frame #0    Kdisplay_conditionYcommand_linesVenvironment_commandsI       0x000000005bd608 libenvironment-commands.dylib at command-line.c:11838
  frame #2    Kanonymous_of_do_execute_commandF235I                        0x00000000008265 libdylan-compiler.dylib at command-line.c:2857
  frame #4    KsearchF64I                                                  0x00000000035a54 libdylan.dylib at condition.c:2794
  frame #5    KsignalVKdMM0I                                               0x00000000034fc4 libdylan.dylib at condition.c:2714
  frame #6    KerrorVKdMM0I                                                0x0000000003525c libdylan.dylib at condition.c:2633
  frame #7    Kelement_range_errorVKeI                                     0x000000000431f1 libdylan.dylib at collection.c:6003
  frame #8    KelementVKdMM3I                                              0x00000000051709 libdylan.dylib at vector.c:2467
  frame #10   Kwrite_instruction_recordYllvm_internalsVllvmMM25I           0x00000001477aa9 libllvm.dylib at llvm-bitcode.c:15467
  frame #11   Kwrite_functionYllvm_internalsVllvmI                         0x00000001483228 libllvm.dylib at llvm-bitcode.c:12426
  frame #12   Kwrite_moduleYllvm_internalsVllvmI                           0x0000000147e3eb libllvm.dylib at llvm-bitcode.c:9148
  frame #13   Kllvm_write_bitcodeYllvm_internalsVllvmI                     0x000000014744c1 libllvm.dylib at llvm-bitcode.c:7843
  frame #15   Kllvm_save_bitcode_fileVllvmI                                0x000000014743a9 libllvm.dylib at llvm-bitcode.c:7794
  frame #17   Kanonymous_of_emit_library_recordF49I                        0x00000001507aa3 libdfmc-llvm-linker.dylib at llvm-linker.c:2376
```

The code involved:

``` dylan
define method write-instruction-record
    (stream :: <bitcode-stream>,
     instruction-index :: <integer>,
     type-partition-table :: <object-table>,
     value-partition-table :: <explicit-key-collection>,
     attributes-index-table :: <encoding-sequence-table>,
     value :: <llvm-call-instruction>)
 => ();
  let attribute-list-encoding
    = encode-attribute-list(value.llvm-call-instruction-attribute-list);
  let operands = make(<stretchy-object-vector>);
  add-value-type(operands, instruction-index,
                 type-partition-table, value-partition-table,
                 value.llvm-instruction-operands[0]);

  // Fixed parameters
  let callee = value-forward(value.llvm-instruction-operands[0]);
  let function-pointer-type
    = type-forward(llvm-value-type(callee));
  let function-type
    = type-forward(function-pointer-type.llvm-pointer-type-pointee);
  let fixed-parameter-count
    = function-type.llvm-function-type-parameter-types.size;
  for (i from 1 below fixed-parameter-count + 1)
    add-value(operands, value-partition-table,
              value.llvm-instruction-operands[i]);
  end for;

  // varargs parameters
  if (function-type.llvm-function-type-varargs?)
    for (i from fixed-parameter-count + 1
           below value.llvm-instruction-operands.size)
      add-value-type(operands, instruction-index,
                     type-partition-table, value-partition-table,
                     value.llvm-instruction-operands[i]);
    end for;
  end if;

  write-record(stream, #"INST_CALL",
               attributes-index-table[attribute-list-encoding],
               logior(ash(value.llvm-call-instruction-calling-convention, 1),
                      if(value.llvm-call-instruction-tail-call?) 1 else 0 end),
               operands);
end method;
```

with the error coming from within:

``` dylan
  for (i from 1 below fixed-parameter-count + 1)
    add-value(operands, value-partition-table,
              value.llvm-instruction-operands[i]);
  end for;
```

In the C, we're at:

``` c
   15465          // opendylan/sources/lib/llvm/llvm-bitcode.dylan:1960
   15466          CONGRUENT_CALL_PROLOG(&KelementVKd, 3);
-> 15467          T20 = CONGRUENT_CALL3(T33, T35, &KPempty_vectorVKi);
```

If we print T33 and T35 here:

```
(lldb) expr -P2 -- T33
(dylan_value) $0 = 0x030d4090 {<simple-object-vector>: size: 2} {
  [0] = 0x0370d540 {<llvm-function>} {
    [llvm-value-type] = 0x037d7fd0 {<llvm-pointer-type>}
    [llvm-global-name] = 0x02620c58 {<byte-string>: size: 24, data: "primitive_wrap_c_pointer"}
    [llvm-global-linkage-kind] = 0x010eb6bc {<symbol>: #"external"}
    [llvm-global-visibility-kind] = 0x00105e38 {<symbol>: #"default"}
    [llvm-global-unnamed-address?] = 0x000e96f0 {<boolean>: #f}
    [llvm-global-alignment] = 0x000e96f0 {<boolean>: #f}
    [llvm-global-section] = 0x000e96f0 {<boolean>: #f}
    [llvm-function-calling-convention] = 0x00000021 {<integer>: 8}
    [llvm-function-garbage-collector] = 0x000e96f0 {<boolean>: #f}
    [llvm-function-arguments] = 0x030d40f0 {<simple-object-vector>: size: 2}
    [llvm-function-attribute-list] = 0x0200a618 {<llvm-attribute-list>}
    [llvm-function-basic-blocks] = 0x037d7fc0 {<stretchy-object-vector>}
    [llvm-function-value-table] = 0x03653020 {<string-table>}
  }
  [1] = 0x03653160 {<llvm-call-instruction>} {
    [llvm-instruction-operands] = 0x030d43d8 {<simple-object-vector>: size: 2}
    [llvm-instruction-metadata] = 0x037d6560 {<pair>}
    [llvm-value-type] = 0x020ec7a0 {<llvm-pointer-type>}
    [llvm-call-instruction-attribute-list] = 0x0200a618 {<llvm-attribute-list>}
    [llvm-call-instruction-calling-convention] = 0x00000001 {<integer>: 0}
    [llvm-call-instruction-tail-call?] = 0x000e96f0 {<boolean>: #f}
  }
}
(lldb) expr -- T35
(dylan_value) $1 = 0x00000009 {<integer>: 2}
```

**Comment by waywardmonkeys** [comment]
It looks like I'm calling ``primitive-wrap-c-pointer`` incorrectly and I don't have a build with type checking for ``primitive-*`` calls enabled. Unclear if this is actually a bug once we fix the type checking...

**Comment by housel** [comment]
I believe this was fixed in 84a4267fbb, which adds function call arity checking within the builder method (`ins--call`).
